### PR TITLE
Resolve #36: 時系列交差検証（TimeSeriesSplit）の導入

### LINKED ISSUE #36: 🟡 [High] 時系列交差検証が未実装（過学習検出が不十分）
## 問題内容

`train_real_marine.py:139` で `shuffle=False` を設定して時系列を崩さないようにしているが、**時系列交差検証（TimeSeriesSplit）を使っていない**ため、未来予測の汎化性能が不明。

## 影響

- 過学習の検出が甘い（たまたまテストデータの期間が予測しやすかった可能性）
- 異なるデータ期間での再現性検証ができない
- モデルの真の予測能力が測定できていない

## 証拠

- `train_real_marine.py:139`: `train_test_split(X, y, test_size=0.2, shuffle=False)`
- コメントに「とりあえずベースライン」と記載あり

### 1. TimeSeriesSplit を導入

```python
from sklearn.model_selection import TimeSeriesSplit

tscv = TimeSeriesSplit(n_splits=5)
scores = []
for train_index, test_index in tscv.split(X):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]
    model.fit(X_train, y_train)
    scores.append(r2_score(y_test, model.predict(X_test)))
print(f"CV平均R²: {np.mean(scores):.4f} ± {np.std(scores):.4f}")
```

### 2. 複数の評価期間でテスト
- 2023年データで学習 → 2024年で評価
- 2022-2023年で学習 → 2024年で評価
- 季節ごとの精度差を確認

### 3. 残差分析
予測誤差が大きい日付を抽出し、共通パターン（台風、黒潮蛇行など）を調査

## 優先度
🟡 **High** - モデル評価の信頼性に関わる

### PR DESCRIPTION
**Pull request by kenta10403**
🤖 **Antigravityによる自動実装・PR作成**

## 概要
- `train_real_marine.py` および `train_d.py` の学習パイプラインに対し、時系列のデータ構造を維持したまま汎化性能を評価するため、`TimeSeriesSplit` (n_splits=5) を導入しました。
- 各モデルの学習前にCV（交差検証）を実行し、R2スコアおよびRMSE (MSE) の平均・標準偏差をログに出力するように改善しています。

これにより、過学習の兆候を早期に検出し、モデルの真の予測能力を適切に評価可能になります。

Closes #36

### CODE REVIEW

**Comment by kenta10403** [comment]
🤖 **Claude Codeによる自動コードレビュー**

## 総合評価: 👍 優れた改善！（ただし PR #52 との調整が必要）

**TimeSeriesSplit** の導入は時系列データのモデル評価として非常に適切な改善や！

---

## ✅ 良い点

1. **適切な交差検証手法の採用**
   - `TimeSeriesSplit` は時系列データの評価に最適
   - 未来データの漏洩を防ぐ正しいアプローチ

2. **過学習検出の強化**
   - CV R2 と Hold-out R2 の両方を表示
   - 標準偏差も計算してモデルの安定性を評価
   - これによって過学習の早期発見が可能に！

3. **コードの品質**
   - `verbosity=-1` でLightGBMの冗長な出力を抑制
   - `reset_index(drop=True)` で安全なCV分割
   - コメントで各ステップの意図が明確

4. **3つのモデル全てに適用**
   - `train_d.py` (Trend Model)
   - `train_real_marine.py` (Marine Env Model)
   - `train_real_marine.py` (Catch Forecast Model)

---

## 🟡 改善提案

### 1. n_splits の妥当性

```python
tscv = TimeSeriesSplit(n_splits=5)
```

`n_splits=5` はええ選択やけど、**データ量によっては調整が必要**かもしれへんで。特に海況モデルは実測値が少ない（8日に1回）から、各foldのサンプル数が少なくなりすぎる可能性があるな。

**推奨**: データ量に応じて `n_splits=3` も検討してみて！

### 2. CV結果の保存

現在、CV結果はコンソールに表示されるだけやけど、**モデルファイルに保存**しておくと後で比較できて便利やで！

```python
model_data = {
    "model": model,
    "features": features,
    "train_means": train_means,
    "cv_r2_mean": np.mean(cv_r2),  # 追加
    "cv_r2_std": np.std(cv_r2)      # 追加
}
```

### 3. 早期停止（Early Stopping）の検討

LightGBMには Early Stopping があるから、CV結果を使って過学習を防ぐこともできるで！（今は n_estimators=100 固定）

---

## 🔴 重大な懸念事項

### PR #52 との競合

**PR #52** も同じファイル（`train_d.py`, `train_real_marine.py`）を変更してるで。

- **PR #51 (このPR)**: TimeSeriesSplit を追加
- **PR #52**: ラグ変数を削除

マージ順序によっては **コンフリクト** が発生する可能性があるな。

### 推奨アクション

1. **PR #52 とこのPRをどちらも適用したい場合**
   - PR #52 を先にマージ
   - このPRを Rebase してから再度レビュー

2. **または両方の変更を1つのPRに統合**
   - TimeSeriesSplit + ラグ変数削除を同時に実装
   - これが一番クリーンなアプローチやと思うで！

---

## 📊 期待される効果

この変更によって：
- ✅ モデルの汎化性能をより正確に評価できる
- ✅ 過学習の早期発見が可能
- ✅ モデルの安定性（標準偏差）を可視化

めっちゃええ改善や！Issue #36 の解決として適切やで！🎣

**Comment by kenta10403** [comment]
PR #52 にTimeSeriesSplitの変更が完全に含まれているため、このPRはクローズします。Issue #36 は PR #52 のマージにより解決済みです。
